ledger/src/bank_forks.rs

```diff
@@ -177,35 +177,50 @@ impl BankForks {
             .last()
             .map(|bank| bank.transaction_count())
             .unwrap_or(0);
-
-        root_bank.squash();
-        let new_tx_count = root_bank.transaction_count();
-
-        // Generate a snapshot if snapshots are configured and it's been an appropriate number
-        // of banks since the last snapshot
+        // Generate each snapshot at a fixed interval
         if self.snapshot_config.is_some() && snapshot_package_sender.is_some() {
             let config = self.snapshot_config.as_ref().unwrap();
-            info!("setting snapshot root: {}", root);
-            if root - self.last_snapshot_slot >= config.snapshot_interval_slots as Slot {
-                let mut snapshot_time = Measure::start("total-snapshot-ms");
-                let r = self.generate_snapshot(
-                    root,
-                    &root_bank.src.roots(),
-                    snapshot_package_sender.as_ref().unwrap(),
-                    snapshot_utils::get_snapshot_archive_path(&config.snapshot_package_output_path),
-                );
-                if r.is_err() {
-                    warn!("Error generating snapshot for bank: {}, err: {:?}", root, r);
-                } else {
-                    self.last_snapshot_slot = root;
+            let mut banks = vec![root_bank];
+            let parents = root_bank.parents();
+            banks.extend(parents.iter());
+            for bank in banks {
+                let parent_slot = bank.parent_slot();
+                let root = bank.slot();
+                if Self::check_interval(root, parent_slot, config.snapshot_interval_slots) {
+                    // Generate a snapshot if snapshots are configured and it's been an appropriate number
+                    // of banks since the last snapshot
+                    info!(
+                        "setting snapshot root: {} interval: {}",
+                        root, config.snapshot_interval_slots
+                    );
+                    if root - self.last_snapshot_slot >= config.snapshot_interval_slots as Slot {
+                        bank.squash();
```

Comment:
it seems that `Bank::squash` isn't idempotent due to its calls of `add_root`, so if `root_bank` == `bank` and `Self::check_interval == true`, we call `squash` on the same bank [twice](https://github.com/solana-labs/solana/pull/8314/files?utf8=%E2%9C%93&diff=unified&w=1#diff-f5031fae39eabfef864ee62e92742a1fR222). I don't think this is good thing to do?

Reply:
good catch!

Reply:
I'm actually not sure this method of looking at the parents which could be already squashed works, because the old rooted banks could have accounts which are cleaned up and then it would not have the same view into the account set as when it was just recently rooted without cleanup possible.